File: clojure/nth-prime/src/nth_prime.clj
(ns nth-prime)

(defn prime? [n]
  (let [r (int (Math/sqrt n))]
    (every? #(not= 0 (mod n %)) (range 5 (inc r)))))

(defn possible-primes [n]
  (lazy-seq (cons (dec n) (cons (inc n) (possible-primes (+ n 6))))))

(def primes
  (cons 2 (cons 3 (filter prime? (possible-primes 6)))))

(defn nth-prime [n]
  (if (zero? n)
    (throw (IllegalArgumentException. "n must be positive"))
    (nth primes (dec n))))

File: r/nth-prime/nth-prime.R
prime <- function(n) {
  stopifnot(n > 0)

  primes <- c(2, 3)
  current <- 6

  while (length(primes) < n) {
    limit <- sqrt(current + 1)
    prime_check <- primes[primes <= limit]

    if (all((current - 1) %% prime_check != 0)) {
      primes <- c(primes, current - 1)
    }

    if (all((current + 1) %% prime_check != 0)) {
      primes <- c(primes, current + 1)
    }

    current <- current + 6
  }

  primes[n]
}
